src/aspire/ctf/estimate_ctf.py

```diff
@@ -0,0 +1,203 @@
+import logging
+import os
+
+import mrcfile
+import numpy as np
+
+from aspire.basis.ffb_2d import FFBBasis2D
+from aspire.ctf import CtfEstimator
+
+logger = logging.getLogger(__name__)
+
+
+def estimate_ctf(
+    data_folder,
+    pixel_size,
+    cs,
+    amplitude_contrast,
+    voltage,
+    num_tapers,
+    psd_size,
+    g_min,
+    g_max,
+    corr,
+    output_dir,
+    dtype=np.float32,
+):
+    """
+    Given paramaters estimates CTF from experimental data
+    and returns CTF as a mrc file.
+    """
+
+    dtype = np.dtype(dtype)
+    assert dtype in (np.float32, np.float64)
+
+    dir_content = os.scandir(data_folder)
+
+    mrc_files = [f.name for f in dir_content if os.path.splitext(f)[1] == ".mrc"]
+    mrcs_files = [f.name for f in dir_content if os.path.splitext(f)[1] == ".mrcs"]
+    file_names = mrc_files + mrcs_files
+
+    amp = amplitude_contrast
+    amplitude_contrast = np.arctan(
+        amplitude_contrast / np.sqrt(1 - np.square(amplitude_contrast))
+    )
+    lmbd = 1.22639 / np.sqrt(voltage * 1000 + 0.97845 * np.square(voltage))
+
+    ctf_object = CtfEstimator(
+        pixel_size, cs, amplitude_contrast, voltage, psd_size, num_tapers, dtype=dtype
+    )
+
+    # Note for repro debugging, suggest use of doubles,
+    #   closer to original code.
+    ffbbasis = FFBBasis2D((psd_size, psd_size), 2, dtype=dtype)
+
+    results = []
+    for name in file_names:
+        with mrcfile.open(
+            os.path.join(data_folder, name), mode="r", permissive=True
+        ) as mrc:
+            micrograph = mrc.data
+
+        # Try to match dtype used in Basis instance
+        micrograph = micrograph.astype(dtype, copy=False)
+
+        micrograph_blocks = ctf_object.ctf_preprocess(micrograph, psd_size)
+
+        tapers_1d = ctf_object.ctf_tapers(
+            psd_size, (2 * num_tapers) / psd_size, num_tapers
+        )
+
+        signal_observed = ctf_object.ctf_estimate_psd(
+            micrograph_blocks, tapers_1d, num_tapers
+        )
+
+        thon_rings, g_out = ctf_object.ctf_elliptical_average(
+            ffbbasis, signal_observed, 0
+        )  # absolute differenceL 10^-14. Relative error: 10^-7
+
+        # Optionally changing to: linprog_method='simplex',
+        # will more deterministically repro results in exchange for speed.
+        signal_1d, background_1d = ctf_object.ctf_background_subtract_1d(
+            thon_rings, linprog_method="interior-point"
+        )
+
+        if corr:
+            avg_defocus, low_freq_skip = ctf_object.ctf_opt1d(
+                signal_1d,
+                pixel_size,
+                cs,
+                lmbd,
+                amplitude_contrast,
+                signal_observed.shape[-1],
+            )
+
+            signal, background_2d = ctf_object.ctf_background_subtract_2d(
+                signal_observed, background_1d, low_freq_skip
+            )
+
+            ratio = ctf_object.ctf_PCA(signal_observed, pixel_size, g_min, g_max)
```

Comment:
Same in both branches.

Reply:
I think i refactored this as part of another suggestion.